subcommand/mesh-init/dataplane-monitor.go

```diff
@@ -0,0 +1,99 @@
+package meshinit
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/hashicorp/consul-ecs/awsutil"
+	"github.com/hashicorp/consul-ecs/config"
+	"github.com/hashicorp/go-hclog"
+)
+
+type dataplaneMonitor struct {
+	ctx context.Context
+	log hclog.Logger
+
+	doneCh chan struct{}
+}
+
+func newDataplaneMonitor(ctx context.Context, logger hclog.Logger) *dataplaneMonitor {
+	return &dataplaneMonitor{
+		ctx:    ctx,
+		log:    logger,
+		doneCh: make(chan struct{}, 1),
+	}
+}
+
+func (d *dataplaneMonitor) done() chan struct{} {
+	return d.doneCh
+}
+
+// Run will wake up when SIGTERM is received. Then, it polls task metadata
+// until the dataplane container stops. Use the Done() channel to wait
+// until it has finished.
+func (d *dataplaneMonitor) run() {
+	defer close(d.doneCh)
+
+	if !d.waitForSIGTERM() {
+		d.doneCh <- struct{}{}
+		return
+	}
+
+	d.log.Info("waiting for dataplane container to stop")
+	for {
+		select {
+		case <-d.ctx.Done():
+			return
+		case <-time.After(1 * time.Second):
+			taskMeta, err := awsutil.ECSTaskMetadata()
+			if err != nil {
+				d.log.Error("fetching task metadata", "err", err.Error())
+				break
```

Comment:
What happens to anything waiting on the `doneCh` if we exit here due to an error getting the `ECSTaskMetadata` or if we return due to `d.ctx.Done()`?

Reply:
I have added a `defer close(doneCh)` at the start of this function. Wouldn't that take of care of the cases you mentioned?

Reply:
Yes I think that covers it. Missed that the first time around.